### DISCUSSION TITLE
How to build libvips with mozjpeg

### DISCUSSION THREAD

**Discussion by rektabhi** [libvips/libvips#3320] [Q&A]
Hi
I am building libvips 8.13 from source on Debian 11. It uses libjpeg as of now. Instead of libjpeg, I want to use mozjpeg. I tried building mozjpeg from source and the build libvips. But somehow, libvips is still using libjpeg.

```
apt-get update && apt-get install -y \
    build-essential \
    ninja-build \
    meson \
    wget \
    unzip \
    cmake \
    nasm \
    pkg-config

apt-get install -y \
    glib-2.0-dev \
    libexpat-dev \
    librsvg2-dev \
    libexif-dev \
    liblcms2-dev \
    libpango1.0-dev \
    libgirepository1.0-dev \
    liborc-dev

tar xf mozjpeg-4.1.1.tar.gz
cd mozjpeg-4.1.1
mkdir build && cd build
sudo cmake -G"Unix Makefiles" ../
sudo make install
cd /usr/local/bin
sudo ln -s /opt/mozjpeg/bin/cjpeg
sudo ln -s /opt/mozjpeg/bin/djpeg
cd -
cd ../..

export PKG_CONFIG_PATH=/opt/mozjpeg/lib64/pkgconfig

VIPS_VERSION=8.13.1

tar xf vips-${VIPS_VERSION}.tar.gz \
&& cd vips-${VIPS_VERSION} \
&& meson build --buildtype=release --libdir=lib \
&& cd build \
&& ninja \
&& ninja install \
&& ldconfig
```

But when I check the output of ldd command, I see libjpeg only.
```
ldd `which vips`  | grep jpeg
libjpeg.so.62 => /lib/x86_64-linux-gnu/libjpeg.so.62 (0x00007f549f282000)
```

Am I doing something wrong in the setup? How do I force libvips to use mozjpeg?

**Comment by jcupitt** [comment]
Hi, there's a sample build here:

https://github.com/jcupitt/docker-builds/blob/master/libvips-mozjpeg-ubuntu20.04/Dockerfile

The problem you're having is that libjpeg is a fundamental library, so it's used everywhere. It's not enough to build libvips against libmozjpeg, you need to also build all direct and indirect dependencies of libvips against libmozjpeg.

**Reply by rektabhi** [reply]
@jcupitt Thanks for the quick response. Are you familiar with any library which might be using `libjpeg`? I checked with `apt show` and could not find any packages in the above script with `libjpeg` dependency.

**Reply by jcupitt** [reply]
I would check rsvg and pango.

**Reply by jcupitt** [reply]
libvips might also be detecting something like imagemagick and linking to that.

**Comment by rektabhi** [comment]
@jcupitt I read that `mozjpeg` could be slower than `libjpeg-turbo` during encoding. `mozjpeg` focusses on reducing the size of the jpeg more than the encoding performance. I wanted something which is extremely fast in encoding, which makes me think `libjpeg-turbo` will be more suited to my usecase. Do you think my understanding is correct?
Also is the installation process similar to mozjpeg? Any sample Dockerfile would be useful.

**Reply by jcupitt** [reply]
Recent mozjpeg includes some code from libjpeg-turbo, I believe, so encoding performance is not too bad. I would benchmark and test.

libjpeg-turbo is usually the default jpeg encode library. Check your package manager. On debian-style systems all you do is `apt install libjpeg-turbo` and it does everything for you.

**Reply by rektabhi** [reply]
Thanks @jcupitt. I was able to install it on Debian 11 using `sudo apt-get install libjpeg62-turbo-dev`. This gave me around 10% improvement in the latencies.

I also wanted to replace libpng with libspng which I was able to do. But the improvement that I see is not that great. I was thinking if I can also replace the default `zlib` library with the `zlib-ng` library. Can you help me with the same? How do I verify if `zlib-ng` is getting used?

**Reply by jcupitt** [reply]
Sorry, I've never tried replacing zlib.

**Reply by rektabhi** [reply]
I was able to install `zlib-ng`, and built `libspng` after that. `libspng` was able to find `zlib-ng` and the performance improvement (in terms of latency) for encoding png images were about 30%.

Thanks for all the help and creating the awesome libvips @jcupitt.
